cli/flox-rust-sdk/src/models/lockfile.rs

```diff
@@ -364,42 +461,63 @@ impl LockedManifestCatalog {
+                    install_id: pkg.install_id,
+                    rel_path: pkg.attr_path,
+                    info: PackageInfo {
+                        description: pkg.description,
+                        broken: pkg.broken,
+                        license: pkg.license,
+                        pname: pkg.pname,
+                        unfree: pkg.unfree,
+                        version: Some(pkg.version),
+                    },
+                    priority: Some(pkg.priority),
+                }),
+                LockedPackage::Flake(_) => None,
             })
             .collect()
     }
 
-    /// Produce a lockfile for a given manifest using the catalog service.
+    /// Produce a lockfile for a given manifest.
+    /// Uses the catalog service to resolve [ManifestPackageDescriptorCatalog],
+    /// and an [InstallableLocker] to lock [ManifestPackageDescriptorFlake] descriptors.
     ///
     /// If a seed lockfile is provided, packages that are already locked
-    /// will constrain the resolution.
+    /// will constrain the resolution of catalog packages to the same derivation.
+    /// Already locked flake installables will not be locked again,
+    /// and copied from the seed lockfile as is.
+    ///
+    /// Catalog and flake installables are locked separately, usinf largely symmetric logic.
+    /// Keeping the locking of each kind separate keeps the existing methods simpler
+    /// and allows for potential parallelization in the future.
     pub async fn lock_manifest(
         manifest: &TypedManifestCatalog,
         seed_lockfile: Option<&LockedManifestCatalog>,
         client: &impl catalog::ClientTrait,
+        installable_locker: &impl InstallableLocker,
     ) -> Result<LockedManifestCatalog, LockedManifestError> {
-        let groups = Self::collect_package_groups(manifest, seed_lockfile)?;
+        let catalog_groups = Self::collect_package_groups(manifest, seed_lockfile)?;
         let (already_locked_packages, groups_to_lock) =
-            Self::split_fully_locked_groups(groups, seed_lockfile);
+            Self::split_fully_locked_groups(catalog_groups, seed_lockfile);
+
+        let flake_installables = Self::collect_flake_installables(manifest);
+        let (already_locked_installables, installables_to_lock) =
+            Self::split_locked_flake_installables(flake_installables, seed_lockfile);
 
         // The manifest could have been edited since locking packages,
         // in which case there may be packages that aren't allowed.
-        Self::check_packages_are_allowed(&already_locked_packages, &manifest.options.allow)?;
+        Self::check_packages_are_allowed(
+            already_locked_packages
+                .iter()
+                .filter_map(LockedPackage::as_catalog_package_ref),
+            &manifest.options.allow,
+        )?;
 
-        if groups_to_lock.is_empty() {
+        if groups_to_lock.is_empty() && installables_to_lock.is_empty() {
```

Comment:
blocking: looks like`already_locked_installables` is getting dropped?

nit: since you're creating empty `vec![]` anyway, why not just get rid of this early return and wrap `check_packages_are_allowed` with this `if` statement?

Reply:
- added `already_locked_installables`
- i think i prefer the early return but if you feel strongly i can change it around

Reply:
Nope that's fine by me